src/doppler-types.ts

```diff
@@ -64,39 +64,210 @@ export interface PrepaidPack {
   name: string;
   credits: number;
   price: number;
+  featureSet: 'standard';
 }
 
 export interface FreePlan {
   type: 'free';
   subscriberLimit: 500;
+  featureSet: 'free';
+}
+
+export interface AgencyPlan {
+  type: 'agency';
+  featureSet: 'agency';
 }
 
 export interface FreePath {
   type: 'free';
   actual: boolean;
-  deadend: true;
+  deadEnd: boolean;
 }
 
-export type Plan = SubscribersLimitedPlan | FreePlan | PrepaidPack | MonthlyRenewalDeliveriesPlan;
+export type Plan =
+  | SubscribersLimitedPlan
+  | FreePlan
+  | PrepaidPack
+  | MonthlyRenewalDeliveriesPlan
+  | AgencyPlan;
 
 export type FeaturedPlan = SubscribersLimitedPlan | MonthlyRenewalDeliveriesPlan;
 
 export interface StandardPath {
   type: 'standard';
   actual: boolean;
   minimumFee: number;
-  deadend: boolean;
+  deadEnd: boolean;
 }
 
 export interface PlusPath {
   type: 'plus';
   actual: boolean;
   minimumFee: number;
-  deadend: boolean;
+  deadEnd: boolean;
 }
 
 export interface AgenciesPath {
   type: 'agencies';
   actual: boolean;
-  deadend: true;
+  deadEnd: boolean;
 }
+
+const getPrepaidPacks = (planList: Plan[]): PrepaidPack[] =>
+  planList.filter((x) => x.type == 'prepaid') as PrepaidPack[];
+
+const getUpgradeMonthlyPlans = (
+  planList: Plan[],
+  { minFee, minEmailsByMonth }: { minFee: number; minEmailsByMonth: number } = {
+    minFee: 0,
+    minEmailsByMonth: 0,
+  },
+): MonthlyRenewalDeliveriesPlan[] =>
+  planList.filter(
+    (x) => x.type == 'monthly-deliveries' && x.fee > minFee && x.emailsByMonth >= minEmailsByMonth,
+  ) as MonthlyRenewalDeliveriesPlan[];
+
+const getUpgradeSubscribersPlans = (
+  planList: Plan[],
+  { minFee, minSubscriberLimit }: { minFee: number; minSubscriberLimit: number } = {
+    minFee: 0,
+    minSubscriberLimit: 0,
+  },
+): SubscribersLimitedPlan[] =>
+  planList.filter(
+    (x) => x.type == 'subscribers' && x.fee > minFee && x.subscriberLimit >= minSubscriberLimit,
+  ) as SubscribersLimitedPlan[];
+
+const getFreePlans = (planList: Plan[]): FreePlan[] =>
+  planList.filter((x) => x.type == 'free') as FreePlan[];
```

Comment:
Here I think returning the plan directly is the best. Same with agency.

Reply:
In my opinion, it allows us to unify the behavior and split the logic into two parts: 

* Defining a list of current and potential upgrade plans.
* Inferring the posible paths based on the previous list.

In my opinion, it also simplifies the logic. Returning the plan directly will require more conditions and the code will be long and complex.